activerecord/test/models/project.rb

```diff
@@ -4,7 +4,7 @@ class Project < ActiveRecord::Base
   has_and_belongs_to_many :non_unique_developers, -> { order 'developers.name desc, developers.id desc' }, :class_name => 'Developer'
   has_and_belongs_to_many :limited_developers, -> { limit 1 }, :class_name => "Developer"
   has_and_belongs_to_many :developers_named_david, -> { where("name = 'David'").distinct }, :class_name => "Developer"
-  has_and_belongs_to_many :developers_named_david_with_hash_conditions, -> { where(:name => 'David').distinct }, :class_name => "Developer"
+  has_and_belongs_to_many :developers_named_david_with_hash_conditions, -> { where(:developers => {:name => 'David'}).distinct }, :class_name => "Developer"
```

Comment:
I don't know if that is correct, however without that change the test I added fails with wrong SQL